Refactor solver.cpp and Core solver creation

delcypher opened the pull request:
This PR attempts to clean up ``Solver.cpp`` and how the backend solver (core solver) is created.

This is in preparation for adding Z3 support to KLEE.

Note that this PR very likely breaks compilation of KLEE with metaSMT support. I don't have metaSMT built on my system (and I'm not interested in trying) and we don't have TravisCI testing for it. If we can't find a maintainer (e.g. @hpalikareva ) for the metaSMT code then I suggest we drop it soon before the code rot gets any worse

delcypher commented:
Okay. This PR now passes the LLVM 2.9 build. @ccadar @MartinNowack any thoughts?

MartinNowack commented on lib/Solver/DummySolver.cpp:
```diff
@@ -0,0 +1,47 @@
+//===-- DummySolver.h -----------------------------------------------------===//
```
DummySolver.cpp

delcypher replied:
Oops. I'll fix that.

MartinNowack commented:
Very nice!! Nice progress and clean up.
One major thing, for me the separation of Solver vs. SolverImpl is to implement the pimpl idiom.
Therefore, Solver*Impl.h can and should be merged with its Solver*Impl.cpp into Solver*.cpp (i.e. STPSolverImpl.h and STPSolverImpl.cpp into STPSolver.cpp). There is no point of separating it into a separate file if you not plan to access it from another file. This way, implementation is just in one file and we have to deal with less files. 

Second, do you plan of moving TimedSolver as well?

Beside that, very nice job!!

delcypher commented:
> Very nice!! Nice progress and clean up.

Thanks.

> One major thing, for me the separation of Solver vs. SolverImpl is to implement the pimpl idiom.
>Therefore, SolverImpl.h can and should be merged with its SolverImpl.cpp into Solver*.cpp (i.e. STPSolverImpl.h and STPSolverImpl.cpp into STPSolver.cpp). There is no point of separating it into a separate file if you not plan to access it from another file. This way, implementation is just in one file and we have to deal with less files.

``SolverImpl.h`` was already in a separate header file before I started this work. It needs to be in a header file because some of the implementations need to be able to call methods on pointers to  ``SolverImpl`` objects

E.g.

```
bool ValidatingSolver::computeTruth(const Query &query, bool &isValid) {
  bool answer;

  if (!solver->impl->computeTruth(query, isValid))
    return false;
  if (!oracle->impl->computeTruth(query, answer))
    return false;

  if (isValid != answer)
    assert(0 && "invalid solver result (computeTruth)");

  return true;
}
```

If ``SolverImpl.h`` is not declared then the call to ``solver->impl->computeTruth()`` will fail to compile because ``impl`` will be an opaque pointer.

However for other files (i.e. ``STPSolverImpl.{h,cpp}``, ``ValidatingSolver.{h,cpp}``, ``DummySolver.{h,cpp}, ``MetaSMTSolverImpl.{h,cpp}``) they don't need to split into separate header and implementation files. So you're right there. I'll fix these.

> Second, do you plan of moving TimedSolver as well?

What is it that you think needs doing to ``TimedSolver``? That lives in ``Core`` and not in the ``Solver`` library so I wasn't planning on touching it. It is also already split into a header file and implementation file.

domainexpert commented:
@delcypher @MartinNowack You are discussing `TimingSolver` and not `TimedSolver` right? `TimingSolver` does not seem to solve constraints (also not inheriting `Solver` class), but instead implements a mechanism to time out other solvers. So perhaps `SolverTimer` is a better name for it, which avoids confusion with `*Solver` classes that actually implement solving.

domainexpert commented:
An option is to allow `TimingSolver` extends `Solver`, however, I'd rather have them separated so as not to make it too easy for anyone to mistakenly wrap a `TimingSolver` within another `TimingSolver` resulting in timing an already timed operation. Another possible alternative name for `TimingSolver` - if it wasn't a `Solver` - would be `TimedSolverRunner`.

delcypher commented:
@domainexpert 
> @delcypher @MartinNowack You are discussing TimingSolver and not TimedSolver right?
Oops yes we are talking about ``TimingSolver``, there is no ``TimedSolver``.

> So perhaps SolverTimer is a better name for it, which avoids confusion with *Solver classes that actually implement solving.

Perhaps but I think changing that is not really related to this PR so I'm going to change it for now.

delcypher commented:
@MartinNowack I think I've followed your suggestions correctly. I've force pushed my changes so the commits are different now. Is there anything you'd like to change?

ccadar commented:
Thanks, @delcypher . I also took a quick look, and this looks good to me.  I'm not sure we need a separate CoreSolver.cpp file, but it's fine.

delcypher commented:
@ccadar @MartinNowack 
I decided to check if this PR built against MetaSMT....

Building MetaSMT is a serious PITA, especially a version from 2 years ago (as per http://srg.doc.ic.ac.uk/projects/klee-multisolver/getting-started.html). After quite a lot of hacking I final managed to built it (I had to patch STP r940 with https://github.com/stp/stp/commit/ece1a55fb367bd905078baca38476e35b4df06c3.patch). I also had to recompile LLVM 3.4 with RTTI :(

I realised that I could make my life a lot easier if I could build KLEE without STP and just MetaSMT so I've added support for this (I was going to add something like this for my Z3 support anyway). Unsurprisingly once I had done this there were a few places where compilation was broken and I've fixed those.

So now it is possible to compile and link KLEE against MetaSMT only. Unfortunately (and not surprisingly) a bunch of the tests and unittests fail. This PR is not responsible for these failures (except the unittests which were never testing MetaSMT support before I fixed that) so I don't consider them to be blocking this PR.

If someone cares about the MetaSMT they can fix these issues but I'm not going to do it.

Provided the TravisCI build passes with my new changes I will merge this PR.

delcypher commented:
Tests pass. Merging.
